ouroboros-network/src/Ouroboros/Network/Protocol/VersionNegotiation/Type.hs

```diff
@@ -0,0 +1,424 @@
+-- unsorted keys (which also means that it will fail if keys are duplicated).
+--
+decodeKnownVersions :: forall ver r s.
+                       ( Ord ver
+                       , Enum ver
+                       , Show ver
+                       , Serialise ver
+                       )
+                    => KnownVersions ver r
+                    -> CBOR.Decoder s (KnownVersions ver r)
+decodeKnownVersions knownVersions = CBOR.decodeMapLen >>= go Nothing []
+  where
+    go !ver0 !rs !l | l <= 0 = return $ KnownVersions $ reverse rs
+                    | otherwise = do
+      ver <- CBOR.decode
+      if Just ver <= ver0
+        then fail "decodeVersionMap: unsorted map"
+        else case findProtocolVersion ver knownVersions of
+          Nothing -> CBOR.decodeTerm >> go (Just ver) rs (pred l)
+          Just pv -> decodeVersionData pv >>= \r -> go (Just ver) (r : rs) (pred l)
+
+-- |
+-- Smart constructor for the sorted list of protocol versions.
+--
+newKnownVersions :: Ord ver => [ProtocolVersion ver r] -> KnownVersions ver r
+newKnownVersions ps = KnownVersions $ sortBy (\(ProtocolVersion verA _) (ProtocolVersion verB _) -> compare verA verB) ps
+
+-- |
+-- Find protocol version in a list of known versions.  It assumes that the
+-- @knownVersion@ list is sorted with respect to @ver@ (which is fullfiled by
+-- @newKnownVersions@ smart constructor).
+--
+findProtocolVersion
+  :: Ord ver
+  => ver
+  -> KnownVersions ver r
+  -> Maybe (ProtocolVersion ver r)
+findProtocolVersion ver (KnownVersions knownVersions) = go (length knownVersions) knownVersions
+    where
+      go _   []   = Nothing
+      go _   [pv] | protocolVersion pv == ver = Just pv
+                  | otherwise                 = Nothing
+      go len pvs  = let l = len `div` 2
+                    in case splitAt l pvs of
+                      (pvs', pv : pvs'') | protocolVersion pv == ver -> Just pv
+                                         | protocolVersion pv >  ver -> go l pvs'
+                                         | protocolVersion pv <  ver -> go (l - 1) pvs''
+                      _                  -> error "findProtocolVersion error"
+
+data ProtocolVersionDescription m where
+       ProtocolVersionDescription :: (Bounded ptcl, Enum ptcl)
+                                  => ptcl
+                                  -> MiniProtocolDescription ptcl m
+                                  -> ProtocolVersionDescription m
+
+-- |
+-- State transitions of the version negotation protocol.
+--
+data VersionNegotiationProtocol ver r where
+    StPropose :: VersionNegotiationProtocol ver r
+    StConfirm :: VersionNegotiationProtocol ver r
+    StDone    :: VersionNegotiationProtocol ver r
+
+
+instance Protocol (VersionNegotiationProtocol ver r) where
+
+    data Message (VersionNegotiationProtocol ver r) from to where
+
+      -- |
+      -- Propose a list of versions together with version data.  It must be
+      -- encoded to a sorted list.
```

Comment:
Misplaced / irrelevant comment. The `KnownVersions` type states that requirement. I bring it up thought because I think it's not ideal. Use a map.